sdk/storage/azblob/blockblob/chunkwriting.go

```diff
@@ -12,225 +12,306 @@ import (
-// TODO(someone): The existing model provides a buffer size and buffer limit as limiting factors.  The buffer size is probably
-// useless other than needing to be above some number, as the network stack is going to hack up the buffer over some size. The
-// max buffers is providing a cap on how much memory we use (by multiplying it times the buffer size) and how many go routines can upload
-// at a time.  I think having a single max memory dial would be more efficient.  We can choose an internal buffer size that works
-// well, 4 MiB or 8 MiB, and auto-scale to as many goroutines within the memory limit. This gives a single dial to tweak and we can
-// choose a max value for the memory setting based on internal transfers within Azure (which will give us the maximum throughput model).
-// We can even provide a utility to dial this number in for customer networks to optimize their copies.
-func copyFromReader(ctx context.Context, from io.Reader, to blockWriter, o UploadStreamOptions) (CommitBlockListResponse, error) {
-	if err := o.format(); err != nil {
+func copyFromReader[T ~[]byte](ctx context.Context, src io.Reader, dst blockWriter, options UploadStreamOptions, getBufferManager func(maxBuffers int, bufferSize int64) bufferManager[T]) (CommitBlockListResponse, error) {
+	options.setDefaults()
+
+	wg := sync.WaitGroup{}       // Used to know when all outgoing blocks have finished processing
+	errCh := make(chan error, 1) // contains the first error encountered during processing
+
+	buffers := getBufferManager(options.Concurrency, options.BlockSize)
+	defer buffers.Free()
+
+	// create initial buffer
+	if _, err := buffers.Grow(); err != nil {
 		return CommitBlockListResponse{}, err
 	}
 
+	// this controls the lifetime of the uploading goroutines.
+	// if an error is encountered, cancel() is called which will terminate all uploads.
+	// NOTE: the ordering is important here.  cancel MUST execute before
+	// cleaning up the buffers so that any uploading goroutines exit first,
+	// releasing their buffers back to the pool for cleanup.
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	var err error
-	generatedUuid, err := uuid.New()
+	// all blocks have IDs that start with a random UUID
+	blockIDPrefix, err := uuid.New()
 	if err != nil {
 		return CommitBlockListResponse{}, err
 	}
-
-	cp := &copier{
-		ctx:    ctx,
-		cancel: cancel,
-		reader: from,
-		to:     to,
-		id:     newID(generatedUuid),
-		o:      o,
-		errCh:  make(chan error, 1),
+	tracker := blockTracker{
+		blockIDPrefix: blockIDPrefix,
+		options:       options,
 	}
 
-	// Send all our chunks until we get an error.
-	for {
-		if err = cp.sendChunk(); err != nil {
+	// This goroutine grabs a buffer, reads from the stream into the buffer,
+	// then creates a goroutine to upload/stage the block.
+	for blockNum := uint32(0); true; blockNum++ {
+		var buffer T
+		select {
+		case buffer = <-buffers.Acquire():
```

Comment:
@JeffreyRichter I must point out that this algorithm will use _up to_ the number of gorutines as specified by `options.Concurrency` but it _does not guarantee_ that this number of goroutines will be used.  In this respect, I think the semantics of `options.Concurrency` are misleading.

Reply:
In fact, the docs are pretty much wrong.
```go
// Concurrency defines the number of concurrent uploads to be performed to upload the file.
// Each concurrent upload will create a buffer of size BlockSize.  The default value is one.
```
If this is the algorithm we want, I'll need to make some adjustments.

Reply:
It would also be interesting to measure both implementations to see how much of a difference there is in terms of throughput.

Reply:
This was originally `MaxConcurrency` however we changed it due to arch board feedback to align with other languages.  I personally have no issues with changing it, I only bring this up for context.

Reply:
I don't feel strongly one way or the other.  I don't think that a customer specifying concurrency expects that it **has to be** used and can't be less.